right/src/usb_report_updater.c

```diff
@@ -57,6 +61,52 @@ mouse_kinetic_state_t MouseScrollState = {
     .acceleratedSpeed = 50,
 };
 
+static void updateOneDirectionSign(int8_t* sign, int8_t expectedSign, uint8_t expectedState, uint8_t otherState) {
+    if (*sign == expectedSign && !activeMouseStates[expectedState]) {
+        *sign = activeMouseStates[otherState] ? -expectedSign : 0;
+    }
+}
+
+// Assume that mouse movement key has been just released. In that case check if there is another key which keeps the state active.
+// If not, check whether the other direction state is active and either flip movement direction or zero the state.
+static void updateDirectionSigns(mouse_kinetic_state_t *kineticState) {
+    updateOneDirectionSign(&kineticState->horizontalStateSign, -1, kineticState->leftState, kineticState->rightState);
+    updateOneDirectionSign(&kineticState->horizontalStateSign,  1, kineticState->rightState, kineticState->leftState);
+    updateOneDirectionSign(&kineticState->verticalStateSign, -1, kineticState->upState, kineticState->downState);
+    updateOneDirectionSign(&kineticState->verticalStateSign,  1, kineticState->downState, kineticState->upState);
+}
+
+// Called on keydown of mouse action. Direction signs ensure that the last pressed action always takes precedence, and therefore
+// have to be updated statefully.
+static void activateDirectionSigns(uint8_t state) {
```

Comment:
This function could be much simpler by not breaking for every case, but gathering 1 and -1 cases.

Reply:
I don't see it. How?

Reply:
It cannot be done, indeed. Sorry, my bad!

Reply:
:-(